### PR TITLE
SQL: Add option to provide the delimiter for the CSV format

### LINKED ISSUE #41634: SQL: Add CSV Separator option
** Current feature : **
Using Elasticsearch 7.0.0
Using SQL with format=csv will return data separated with comma ","
When expected separator on French OS will likely expect ";" meaning the CSV file won't open properly

** Requested feature **:
Ability to specify the separator to use for example : ?format=csv&csvSeparator=%3B

Example : 
```
POST /_sql?format=csv
{
    "query": "SELECT * FROM test"
} 

#this returns this which won't work if the separator is not a coma: 
user,value
John Smith,1
Bob ; Smith,2

#we would want " POST /_sql?format=csv&csvSeparator=%3B " which should return
user;value
John Smith;1
"Bob ; Smith";2

```

### PR DESCRIPTION
**Pull request by bpintea**
This PR adds the option to provide the desired character as the separator
for the CSV format (the default remains comma).
A set of characters are excluded though - like CR, LF, `"` - to avoid
slipping onto the CSV-dialects slope. The tab is also forbidden, the
user needs to choose the `tsv` format explicitely.

Closes #41634.

### CODE REVIEW

**Comment by bpintea** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java [hunk 1]
```diff
@@ -109,15 +109,33 @@ protected String eol() {
...
-        protected String delimiter() {
-            return ",";
+        protected Character delimiter() {
+            return delimiter;
+        }
+
+        @Override
+        protected void setDelimiter(Character delimiter) {
+            if (delimiter == null) {
+                throw new UnsupportedOperationException();
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:122]: I'd say IllegalArgumentException fits better.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java [hunk 2]
```diff
@@ -90,7 +92,25 @@ public void testCsvFormatWithRegularData() {
         assertEquals("string,number\r\n" +
                 "Along The River Bank,708\r\n" +
                 "Mind Train,280\r\n",
-                text);
+            text);
+    }
+
+    public void testCsvFormatWithCustomDelimiterRegularData() {
+        List<Character> forbidden = Arrays.asList('"', '\r', '\n', '\t');
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java:99]: Minor, a set would be faster.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java [hunk 3]
```diff
@@ -124,6 +128,11 @@ public RestResponse buildResponse(SqlQueryResponse response) throws Exception {
         });
     }
 
+    @Override
+    protected Set<String> responseParams() {
+        return textFormat == TextFormat.CSV ? Collections.singleton("delimiter") : Collections.emptySet();
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:133]: Can we use some static variable for "delimiter"?

**Review by matriv — COMMENTED** [review]
Looks good, left a few comments, but additionally should something be adjusted in docs?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java]: Makes sense.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java]: Changed, thanks.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java (hunk 3 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java]: Sure. I've updated along a few other embedded strings, for this change to make sense.

**Comment by bpintea** [comment]
> should something be adjusted in docs?

Of course, missed that, thanks for pointing it out.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java [hunk 4]
```diff
@@ -235,7 +247,7 @@ String contentType(RestRequest request) {
         }
 
         @Override
-        String maybeEscape(String value) {
+        String maybeEscape(String value, Character __) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:250]: Why did you add that second argument? I don't see it being used, or am I missing something :) ?

File: docs/reference/sql/endpoints/rest.asciidoc [hunk 5]
```diff
@@ -106,6 +106,21 @@ s|Description
...
++
+Controls if the CSV header should be included in the response or not. It can take one of the following values:
++
+   - `present` (default): include the header;
+   - `absent`: exclude the header.
++
+* `delimiter` (default: `,`)
++
+Indicates which character should be used to separate the CSV values. The following values cannot be used for this attribute:
+apostrophe (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
```
Comment (matriv) [inline docs/reference/sql/endpoints/rest.asciidoc:121]: Isn't it better to say `double quote`  instead of `apostrophe` ?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java [hunk 6]
```diff
@@ -186,23 +193,28 @@ String maybeEscape(String value) {
...
+                            if (param.toLowerCase(Locale.ROOT).equals(URL_PARAM_HEADER + "=" + PARAM_HEADER_ABSENT)) {
                                 return false;
                             }
                         }
                     }
                 }
                 return true;
             } else {
-                return !header.toLowerCase(Locale.ROOT).equals("absent");
+                List<String> allowed = List.of(PARAM_HEADER_ABSENT, PARAM_HEADER_PRESENT);
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:212]: You could also use Set here, but personally I would use a `switch` statement on the `head.toLowerCase(Locale.ROOT)` to return true, false, and default => throw the exception.

**Review by matriv — APPROVED** [review]
LGTM. Left a few more minor comments.

**Comment by bpintea** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java (hunk 4 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:253]: The second param of `maybeEscape`, the delimiter, is only used for the `CSV` value, but not for the `TSV`.

File: docs/reference/sql/endpoints/rest.asciidoc (hunk 5 above)
Comment (bpintea) [reply docs/reference/sql/endpoints/rest.asciidoc]: Oops, you're right, thanks. Fixed.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java (hunk 6 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java]: Good suggestion, thanks. Fixed.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:253]: Of course, sorry, I was just confused.

File: docs/reference/sql/endpoints/rest.asciidoc [hunk 7]
```diff
@@ -118,7 +118,7 @@ Controls if the CSV header should be included in the response or not. It can tak
 * `delimiter` (default: `,`)
 +
 Indicates which character should be used to separate the CSV values. The following values cannot be used for this attribute:
-apostrophe (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
+double quotes (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
```
Comment (matriv) [inline docs/reference/sql/endpoints/rest.asciidoc:121]:
```suggestion
double quote (`"`), carriage-return (`\r`) and new-line (`\n`). The tab (`\t`) can also not be used, the `tsv` format needs to be used
```
since all the rest are also singular.

**Review by matriv — APPROVED** [review]
LGTM, Thx!
Left one more minor comment in docs.

**Review by astefan — COMMENTED** [review]
LGTM in general, but I have doubts about adding the `header` request parameter.

File: docs/reference/sql/endpoints/rest.asciidoc [hunk 8]
```diff
@@ -106,6 +106,21 @@ s|Description
...
 |===
 
+The `CSV` format accepts the following additional formatting URL query attributes:
+
+* `header`
++
+Controls if the CSV header should be included in the response or not. It can take one of the following values:
++
+   - `present` (default): include the header;
+   - `absent`: exclude the header.
```
[the remark this answers is not in the document]
Comment (bpintea) [reply docs/reference/sql/endpoints/rest.asciidoc:116]: > Why not using a true/false kind of approach with the csv header?

The `absent`/`present` values [were coded in before](https://github.com/elastic/elasticsearch/blob/74c9e5673561eed60700edc5043c5618dad26b2b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L186). And my guess is that the values have been chosen to be in-line with the [CT header values](https://github.com/elastic/elasticsearch/blob/74c9e5673561eed60700edc5043c5618dad26b2b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L136).

> it was a way of distinguishing between the first page of the response and the subsequent pages of it

I see. ..actually how would a subsequent page be requested with a URL attribute in CSV format? I think we have no `OFFSET`, neither at API, nor at SQL level, do we?
I also found no tests covering this scenario or the use of the `header` attribute.

And these two points above...

> why did you expose this request parameter in the first place?

...let me to believe that it might be an intended feature, but no[t / longer] correctly functioning.

And I think it would be good to have it, CSV handling applications generally offer such functionality. So irrespective of the original intent, is there any concern exposing it?

> Also, the initial request this PR is fixing refers only to the delimiter itself.

True. Would it make sense to split the PR (vs. maybe extending this PR's description)?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java [hunk 9]
```diff
@@ -162,52 +190,58 @@ String maybeEscape(String value) {
...
                 value = sb.toString();
             }
+
             return value;
         }
 
         @Override
-        boolean hasHeader(RestRequest request) {
-            String header = request.param("header");
+        boolean needsHeader(RestRequest request) {
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:198]: My interpretation of [what the code does](https://github.com/elastic/elasticsearch/blob/74c9e5673561eed60700edc5043c5618dad26b2b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L169) is that it decides if a header is required in the response and not if the RestRequest has a `header` value at all (either URL attribute or HTTP header value). I.e. for the configuration `header=absent` the function will still return `true`.
But I'm happy to revert or apply a more suggestive name if this be an incorrect interpretation.

File: docs/reference/sql/endpoints/rest.asciidoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (astefan) [reply docs/reference/sql/endpoints/rest.asciidoc:116]: > ...let me to believe that it might be an intended feature, but no[t / longer] correctly functioning.
My 2c on this: if it's requested, we can discuss about adding it. If there is no demand, we don't complicate things.

> I see. ..actually how would a subsequent page be requested with a URL attribute in CSV format? I think we have no OFFSET, neither at API, nor at SQL level, do we?
The first response that comes from ES contains a header called `cursor` to be used for a subsequent call to get the next page of rows. The URL remains the same, but the second call doesn't have a `query` in it, but a `cursor`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:198]: The way I looked at it is from the request perspective: the request has a header or not, thus the formatting behavior changes based on this. I would argue further and say that the response needs a header (thus a `needsHeader` method should belong to the SqlQueryResponse object) and not the row-by-row formatting behavior.

File: docs/reference/sql/endpoints/rest.asciidoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (bpintea) [reply docs/reference/sql/endpoints/rest.asciidoc:116]: > My 2c on this: if it's requested, we can discuss about adding it. If there is no demand, we don't complicate things.

Sure, I can take it out. But just to understand how to take it out:

> it was a way of distinguishing between the first page of the response and the subsequent pages of it

How is this distinguishing being done, since I believe there's no way to specify a `header` URL attribute currently, since it's not "consumed" by the `RestSqlQueryAction` and so using it would generate a `"request [/_sql] contains unrecognized parameter: [header]"` error. Should I remove it entirely then? Sorry if missing something.

> The first response that comes from ES contains a header called cursor to be used for a subsequent call to get the next page of rows.

Right, it's inserted by [`RestSqlQueryAction`](https://github.com/elastic/elasticsearch/blob/380ff0b0d44c0e0036a8934d2b146ee08645ba26/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java#L117), should have spotted it before. Thanks.
Is this documented anywhere though? I couldn't find it on the [REST formats page](https://www.elastic.co/guide/en/elasticsearch/reference/master/sql-rest-format.html). I can add it if not.

[the remark this answers is not in the document]
Comment (astefan) [reply docs/reference/sql/endpoints/rest.asciidoc:116]: > Sure, I can take it out. But just to understand how to take it out:

Just do not expose it as a request parameter. Leave the rest of the logic unchanged and this PR should only address the delimiter part.

> How is this distinguishing being done, since I believe there's no way to specify a header URL attribute currently, since it's not "consumed" by the RestSqlQueryAction and so using it would generate a "request [/_sql] contains unrecognized parameter: [header]" error. Should I remove it entirely then? Sorry if missing something.

I don't think it's actually being used, indeed. I see logic that's based on the presence of the cursor or not, not much about the `header` being present or not in the request...

> Right, it's inserted by RestSqlQueryAction, should have spotted it before. Thanks.
Is this documented anywhere though? I couldn't find it on the REST formats page. I can add it if not.

It's not documented for other formats (except json and txt), either. I think a simple statement change in [this page](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-pagination.html) also mentioning csv and tsv should be enough (ie, no examples).

[the remark this answers is not in the document]
Comment (bpintea) [reply docs/reference/sql/endpoints/rest.asciidoc]: I've reverted the changes here and opened #60271, since this is clearly a (very minor) bug.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java]: > thus a `needsHeader` method should belong to the SqlQueryResponse object

Agreed.

I've reverted the change anyways.

**Comment by bpintea** [comment]
@elasticmachine update branch

**Comment by bpintea** [comment]
@elasticmachine update branch

**Comment by bpintea** [comment]
@elasticmachine update branch

**Review by astefan — APPROVED** [review]
LGTM
